.github/workflows/build.yml

```diff
@@ -57,4 +65,26 @@ jobs:
         run: cargo hack doc --verbose --no-deps --each-feature --no-dev-deps --optional-deps url -p rumqttc -p rumqttd
 
       - name: Test rumqttc and rumqttd
-        run: cargo hack test --verbose --each-feature --optional-deps url -p rumqttc -p rumqttd
+        run: |
+            cargo hack --each-feature --optional-deps url llvm-cov -p rumqttc -p rumqttd --no-report nextest
+
+      - name: Combine coverage
+        run: cargo llvm-cov report --lcov --output-path coverage.lcov
+
+      - name: Upload Coverage Report
+        uses: coverallsapp/github-action@v2.2.3
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          file: ./coverage.lcov
+          parallel: true
+          flag-name: run-${{ matrix.os }}-cargo
```

Comment:
why do we need to upload it to coverallapp?

how would the flow look like? I mean, let's say I open a PR, then if I have to check the coverage, I would need to visit / lookup where on coverallapp is it available? so what is the benefit?

why can't we just comment it right below the PR? and why coverall instead of Codecov?

Reply:
If we install coveralls app, we will get a comment after every run of the test.

Something like this: https://github.com/arunanshub/pyflocker/pull/444#issuecomment-1951575395

Coveralls instead of codecov because it has less failure rates (due to network errors, etc) that codecov. And coveralls especially because our use case is simple: we just need the coverage report and a badge.

Reply:
> If we install coveralls app, we will get a comment after every run of the test.
>
ah okie, I didn't see any comment on this PR, so thought coverall doesn't support it.

> Coveralls instead of codecov because it has less failure rates (due to network errors, etc) that codecov.
>
nice :)

Reply:
> ah okie, I didn't see any comment on this PR, so thought coverall doesn't support it.

We need admin access to install coveralls app.

Reply:
btw, instead of _uploading_ it to somewhere, we can just use some action to add comment in PR?

using actions like https://github.com/thollander/actions-comment-pull-request or https://github.com/marketplace/actions/add-pr-comment#proxy-for-fork-based-prs should do the job.

this way we won't need any other external setup.

downside? we might not be able to track the history, but code coverage history isn't required anyways right?

Reply:
> downside? we might not be able to track the history, but code coverage history isn't required anyways right?

It is beneficial to track how coverage has changed over commits, and comparing it over pull request.

Reply:
> It is beneficial to track how coverage has changed over commits
>
care to explain how is it beneficial?

> comparing pull request coverage to base branch's counterpart
>
okie :+1:

Reply:
> care to explain how is it beneficial?

- Identify regressions: Catch errors introduced unintentionally by seeing drops in covered code.
- Measure test effectiveness: Monitor if new tests cover new code and if existing tests stay relevant.
- Target testing efforts: Prioritize tests for areas with low or decreasing coverage.